### PR TITLE
Group Screens for Web

### PR DESCRIPTION
**Pull request by sjmcculloch**
- New component for Tag selection by single or multi
- New message component for displaying standard messages
- View group screen with filters for location, capability, text and my filters
- Edit screen allowing name, location, capabilities and multiple users at once
- Extended server implementation to handle tags/users on group add/edit
- GroupTable, ScheduleTable components
- Centralised styles to AppStyle.js
- New selectors for groups

image

image

### CODE REVIEW

File: web/src/screens/private/Groups/EditGroup.js [hunk 1]
```diff
@@ -3,39 +3,328 @@ import PropTypes from 'prop-types';
...
+      this.setInitialState(nextProps);
+    }
+  }
+
+  setInitialState(props) {
+    const group = props.user.organisation.groups.find(
+      g => g.id === parseInt(props.match.params.id, 10),
+    );
+    if (group !== undefined) {
+      let regions = '';
```
Comment (OSPFNeighbour) [inline web/src/screens/private/Groups/EditGroup.js]: do we have a plan on how to handle the different name we might want to call this? org_structure vs regions vs zone vs units etc etc?

**Comment by OSPFNeighbour** [comment]
That list of users will get pretty crazy with a few thousand in it

File: server/src/logic.js [hunk 2]
```diff
@@ -6,33 +6,27 @@ import { JWT_SECRET } from './config';
...
 // reusable function to check for a user with context
-const getAuthenticatedUser = ctx => ctx.user.then((user) => {
-  // null means we couldn't find the user record
-  if (!user) {
-    // eslint-disable-next-line prefer-promise-reject-errors
-    return Promise.reject('Unauthorized: Invalid user ID');
-  }
-  return user;
-});
+const getAuthenticatedUser = ctx =>
```
Comment (sdunster) [inline server/src/logic.js]: again with the whitespace changes making it really hard to review.. if you're going to do this stuff I'd really appreciate putting it in a separate commit so I can review it separately.

Comment (sdunster) [reply server/src/logic.js]: ok I got about halfway through and gave up. Can you comment on the changes?

File: web/src/screens/private/Events/EditEvent.js [hunk 3]
```diff
@@ -246,172 +254,167 @@ class EditEvent extends React.Component {
...
-                      activeStep !== 0 &&
-                      this.state.groupId !== 0 &&
-                      <span> : {this.state.groupName}</span>
-                    }
-                    {index === 1 && activeStep !== 1 && this.state.name !== '' && <span> : {this.state.name}</span>}
-                  </StepLabel>
-                  <StepContent>
-                    <Typography>{getStepContent(index)}</Typography>
-                    {index === 0 && (
+      <div className={classes.root}>
```
Comment (sdunster) [inline web/src/screens/private/Events/EditEvent.js]: this is really really long and probably should be broken up into smaller components

File: web/src/screens/private/Groups/EditGroup.js (hunk 1 above)
Comment (sdunster) [reply web/src/screens/private/Groups/EditGroup.js]: yeah I thought the whole point of this was that we didn't need to hard code these kinds of things :(
Don't want to block progress at this stage but think about it some more if you can.

File: web/src/screens/private/Groups/ViewGroups.js [hunk 4]
```diff
@@ -3,149 +3,198 @@ import PropTypes from 'prop-types';
...
+              justifyContent: 'space-between',
+              alignItems: 'flex-end',
+            }}
+          >
+            <FormControlLabel
+              control={
+                <Switch
+                  checked={this.state.myGroups}
+                  value="myGroups"
+                  onChange={this.handleSwitchChange('myGroups')}
```
Comment (sdunster) [inline web/src/screens/private/Groups/ViewGroups.js]: i don't think you want to call the function as part of the render function here

File: web/src/screens/private/Groups/ViewGroups.js [hunk 5]
```diff
@@ -3,149 +3,198 @@ import PropTypes from 'prop-types';
...
+                />
+              }
+              label="Only My groups"
+            />
+            <div>
+              <FormControl className={classes.formControlFilter}>
+                <Tag
+                  list={locations}
+                  placeholder="Select Location"
+                  onChange={this.handleTagChange('locationFilter')}
```
Comment (sdunster) [inline web/src/screens/private/Groups/ViewGroups.js]: as above

File: web/src/screens/private/Groups/ViewGroups.js [hunk 6]
```diff
@@ -3,149 +3,198 @@ import PropTypes from 'prop-types';
...
+                  value={this.state.locationFilter}
+                  multi={false}
+                />
+              </FormControl>
+              <FormControl className={classes.formControlFilter}>
+                <Tag
+                  list={capabilities}
+                  placeholder="Select Capability"
+                  onChange={this.handleTagChange('capabilityFilter')}
```
Comment (sdunster) [inline web/src/screens/private/Groups/ViewGroups.js]: same

File: web/src/selectors/groups.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+export default (
```
Comment (sdunster) [inline web/src/selectors/groups.js]: if we're not going to name this function can you add a comment explaining what it does?

File: web/src/screens/private/Groups/EditGroup.js [hunk 8]
```diff
@@ -3,39 +3,328 @@ import PropTypes from 'prop-types';
...
+              margin="normal"
+              placeholder="Enter Group..."
+            />
+          </FormControl>
+          <FormControl className={classes.formControl}>
+            <Tag
+              list={regions}
+              label="Regions"
+              placeholder="Select..."
+              onChange={this.handleTagChange('regions')}
```
Comment (sdunster) [inline web/src/screens/private/Groups/EditGroup.js]: dont call function in render

File: web/src/screens/private/Groups/EditGroup.js [hunk 9]
```diff
@@ -3,39 +3,328 @@ import PropTypes from 'prop-types';
...
+              value={this.state.regions}
+              multi
+            />
+          </FormControl>
+          <FormControl className={classes.formControl}>
+            <Tag
+              list={capabilities}
+              label="Capabilities"
+              placeholder="Select..."
+              onChange={this.handleTagChange('capabilities')}
```
Comment (sdunster) [inline web/src/screens/private/Groups/EditGroup.js]: as above

File: web/src/screens/private/Groups/EditGroup.js [hunk 10]
```diff
@@ -3,39 +3,328 @@ import PropTypes from 'prop-types';
...
+              value={this.state.capabilities}
+              multi
+            />
+          </FormControl>
+          <FormControl className={classes.formControl}>
+            <Tag
+              list={users}
+              label="Users"
+              placeholder="Select..."
+              onChange={this.handleTagChange('users')}
```
Comment (sdunster) [inline web/src/screens/private/Groups/EditGroup.js]: as above

File: server/src/logic.js (hunk 2 above)
Comment (sjmcculloch) [reply server/src/logic.js]: whitespaces? I have some auto-formatting in vscode if that's what is causing the issue

File: web/src/screens/private/Events/EditEvent.js (hunk 3 above)
Comment (sjmcculloch) [reply web/src/screens/private/Events/EditEvent.js]: I agree - I was just matching styles and not really touching this area

File: web/src/screens/private/Groups/EditGroup.js (hunk 8 above)
Comment (sjmcculloch) [reply web/src/screens/private/Groups/EditGroup.js]: e.g. I did similar things in SignUpPage.js etc a long time ago 


            <TextField
              required
              fullWidth
              id="password"
              label="Password"
              type="password"
              margin="normal"
              value={this.state.password}
              onChange={this.onPasswordChange}
            />

File: web/src/screens/private/Groups/EditGroup.js (hunk 1 above)
Comment (sjmcculloch) [reply web/src/screens/private/Groups/EditGroup.js]: I'm getting a ton of pressure to get this functionality in, I can move it to config in the future but under time pressure at the moment to get these done.

File: web/src/screens/private/Groups/EditGroup.js (hunk 8 above)
Comment (sjmcculloch) [reply web/src/screens/private/Groups/EditGroup.js]: nevermind - worked out the issue in the react docs

File: server/src/logic.js (hunk 2 above)
Comment (sdunster) [reply server/src/logic.js]: this whole file has changed so much I couldn't be bothered to pick out the actual (non-whitespace) changes. Can you highlight them with inline comments?

File: web/src/screens/private/Groups/ViewGroups.js (hunk 4 above)
Comment (sjmcculloch) [reply web/src/screens/private/Groups/ViewGroups.js]: For hardcoding - the plan is to use config at an org level for what tags are enabled, I can move this at a later date - trying to balance delivering something and keeping it generic at the moment.

**Comment by sjmcculloch** [comment]
Fixing the function references in next PR.

**Pull request merged**